Mathlib/Combinatorics/HalesJewett.lean

```diff
@@ -88,16 +180,61 @@ structure Line (α ι : Type*) where
 #align combinatorics.line Combinatorics.Line
 
 namespace Line
+variable {l : Line α ι} {i : ι} {a x : α}
+
+/-- Consider a line `l : Line α ι` as a function `α → ι → α`. -/
+@[coe] def toFun (l : Line α ι) (x : α) (i : ι) : α := (l.idxFun i).getD x
 
 -- This lets us treat a line `l : Line α ι` as a function `α → ι → α`.
-instance (α ι) : CoeFun (Line α ι) fun _ => α → ι → α :=
+instance instCoeFun : CoeFun (Line α ι) fun _ => α → ι → α :=
   ⟨fun l x i => (l.idxFun i).getD x⟩
 
+lemma coe_apply (l : Line α ι) (x : α) (i : ι) : l x i = (l.idxFun i).getD x := rfl
+
+-- Note: This is not made a `FunLike` instance to avoid having two syntactically different coercions
+lemma coe_injective [Nontrivial α] : Injective ((⇑) : Line α ι → α → ι → α) := by
+  rintro l m hlm
+  ext i a
+  obtain ⟨b, hba⟩ := exists_ne a
+  simp only [Option.mem_def, funext_iff] at hlm ⊢
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · cases hi : idxFun m i <;> simpa [@eq_comm _ a, hi, h, hba] using hlm b i
+  · cases hi : idxFun l i <;> simpa [@eq_comm _ a, hi, h, hba] using hlm b i
+
 /-- A line is monochromatic if all its points are the same color. -/
 def IsMono {α ι κ} (C : (ι → α) → κ) (l : Line α ι) : Prop :=
```

Comment:
IsMonochromatic would be better, because `mono` is often used for `monotone`.

Reply:
Maybe it's worth renaming, but that's out of scope for this PR. `Combinatorics.Line.IsMono` already exists.